pkg/loop/internal/relayer/pluginprovider/chainreader/chain_reader_test.go

```diff
@@ -524,11 +514,57 @@ func (f *fakeChainReader) QueryKey(_ context.Context, _ string, filter query.Key
 			})
 		}
 
-		return sequences, nil
+		// simulate contract being bound at a certain block not having the events before that block unless Replay is called.
+		boundAt, ok := f.contractBoundAt[contractName]
+		if !ok {
+			return nil, fmt.Errorf("contract %s not bound", contractName)
+		}
+
+		if boundAt == 0 {
+			return sequences, nil
+		}
+
+		return sequences[:boundAt], nil
 	}
 	return nil, nil
 }
 
+func (f *fakeChainReader) ReplaySequence(_ context.Context, contractName, _ string, blockID string) error {
+	blockNumber, err := strconv.Atoi(blockID)
+	if err != nil {
+		return err
+	}
+
+	if blockNumber > len(f.triggers) {
```

Comment:
why no read lock?

Reply:
Tests are not parallel, only encoding tests are, but they don't reuse the tester instance